open-craft/opencraft
# [BB-2217] Store static templates customizations and deploy the changes

This PR implements the changes to store static the templates customizations and deploy them by generating the appropriate ansible variables to make use of the changes in https://github.com/open-craft/configuration/pull/121. It also implements the registration v2 API changes to allow setting and patching these values from the frontend.

**Testing instructions**:
* Checkout the source branch of this PR.
* Apply the migrations.
* Start the Ocim app server (`make run.dev` in the devstack).
* Log in to Ocim as a superuser after creating a superuser account, if there isn't one already.
* In a separate terminal tab/window, start the frontend app.
* Register for a new betatest instance using the frontend and verify the email addresses given.
* Once the instance is created automatically after verification, navigate to the swagger API page `http://localhost:5000/api/swagger`.
* Check the listed API endpoints and verify that the `openedx_config` endpoints include the `draft_static_content_overrides` field.
* Verify that in the response of the `/v2/instances/openedx_config/` endpoint, the instance created from the new betatest application is listed and has the `draft_static_templates_overrides` set to `null`, the default value.
* Verify that there is a separate API endpoint which allows the `PATCH` request to modify individual fields.
* In the Django admin interface, navigate to the betatest instance created before and change the `draft_static_content_overrides` field's value to `{"version": 0}` and save it.
* Try executing `PATCH` requests to add more supported fields and update existing fields.
* Trigger a redeployment by executing the `/v2/instances/openedx_deployment/` endpoint with a POST request.
* Verify that the `draft_static_content_overrrides` value is copied to the `static_content_overrides` field in the `OpenEdXInstance` model instance corresponding to the betatest form.
* If it is possible to launch an appserver, verify that there is a section for the `SiteConfiguration` settings in the appserver configuration section. Also verify that the ansible variables there correspond to the format expected by https://github.com/open-craft/configuration/pull/121.
* If it is not possible to launch an appserver in the devstack, run the `.get_site_configuration_settings()` method on the `OpenEdXInstance` model instance and verify the generated ansible variables.
* Try testing other permutations/combinations.

**Author notes and concerns**:
* The `draft_static_content_overrides` field and the `static_content_overrides` field use `None` as their default value. However, to be able to update the former using `PATCH` request, a non-null value for the field must be set. So the expectation is that, whenever the user enables the customization for the first static template, a `PUT` request to the instance config endpoint will be required to populate the field with a non-null value.
* The `create_or_update_site_configuration` management command does not allow deleting any `SiteConfiguration` value keys. So the keys once added, cannot be deleted. However, setting their value to `''` or `null` will trigger the default value in the template to be used.
* I have not updated the frontend API client in this PR as that will add an additional field to the instance config API models and that may or may not break the frontend tests, fixing which is not in the scope of this task. But when the frontend task is being worked upon, I will fix any reported issues with the generated schema.
